cloudslang-compiler/src/main/java/io/cloudslang/lang/compiler/SlangCompilerImpl.java

```diff
@@ -54,20 +56,31 @@
     @Autowired
     private ScoreCompiler scoreCompiler;
 
+    @Autowired
+    private CompileValidator compileValidator;
+
     @Override
-    public CompilationArtifact compile(SlangSource source, Set<SlangSource> path) {
+    public CompilationArtifact compile(SlangSource source, Set<SlangSource> dependencySources) {
 
         Executable executable = preCompile(source);
 
-        //we transform also all of the files in the given path to model objects
-        Set<Executable> pathExecutables = new HashSet<>();
-        if (CollectionUtils.isNotEmpty(path)) {
-            for (SlangSource pathSource : path) {
-                pathExecutables.add(preCompile(pathSource));
+        // we transform also all of the files in the given dependency sources to model objects
+        Set<Executable> allExecutables = new HashSet<>();
+        allExecutables.add(executable);
+
+        if (CollectionUtils.isNotEmpty(dependencySources)) {
+            for (SlangSource currentSource : dependencySources) {
+                Executable preCompiledCurrentSource = preCompile(currentSource);
+
+                compileValidator.validateNoDuplicateExecutablesBasedOnFQN(preCompiledCurrentSource, currentSource, allExecutables);
+
+                allExecutables.add(preCompiledCurrentSource);
             }
         }
 
-        return scoreCompiler.compile(executable, pathExecutables);
+        Set<Executable> dependencies = new HashSet<>(allExecutables);
```

Comment:
just a suggestion, maybe if you renamed allExecutables to executables than you would not need to create a new Set from which you remove the primary executable and remove it directly from the executables Set. WDYT?